### DISCUSSION TITLE
Doubt about Mellinger controller.

### DISCUSSION THREAD

**Discussion by Wallat1** [bitcraze/discussions#952] [Q&A]
Hi.

I have the following doubt, if I am using a controller (xbox or ps), in this part of Mellinger controller

```
ew.x = radians(setpoint->attitudeRate.roll) - stateAttitudeRateRoll;
  ew.y = -radians(setpoint->attitudeRate.pitch) - stateAttitudeRatePitch;
  ew.z = radians(setpoint->attitudeRate.yaw) - stateAttitudeRateYaw;
  if (self->prev_omega_roll == self->prev_omega_roll) { /*d part initialized*/
    err_d_roll = ((radians(setpoint->attitudeRate.roll) - self->prev_setpoint_omega_roll) - (stateAttitudeRateRoll - self->prev_omega_roll)) / dt;
    err_d_pitch = (-(radians(setpoint->attitudeRate.pitch) - self->prev_setpoint_omega_pitch) - (stateAttitudeRatePitch - self->prev_omega_pitch)) / dt;
  }
  self->prev_omega_roll = stateAttitudeRateRoll;
  self->prev_omega_pitch = stateAttitudeRatePitch;
  self->prev_setpoint_omega_roll = radians(setpoint->attitudeRate.roll);
  self->prev_setpoint_omega_pitch = radians(setpoint->attitudeRate.pitch);
```
what values do these terms "setpoint->attitudeRate.roll 
setpoint->attitudeRate.pitch and setpoint->attitudeRate.yaw" take?, since as far as I know, this term does not receive anything when I use a controller.

I would appreciate your help.

Regards.

**Comment by krichardsson** [comment]
Hi,

first of all the mellinger controller is not normally used for manual flight and unless you have configured the Crazyflie to use mellinger, the PID controller is used instead.

Assuming that you have configured the system to use mellinger, I will explain the information flow to enable you to dig into it deeper if you want to.

Setpoints are sent [from the python client](https://github.com/bitcraze/crazyflie-clients-python/blob/master/src/cfclient/ui/main.py#L198) using the [`send_setpoint()` function in the python lib](https://github.com/bitcraze/crazyflie-lib-python/blob/f35b0baf6ff243d618f7a8becf5a69b1ec6c1f56/cflib/crazyflie/commander.py#L70).

As stated in the documentation for the send_setpoint() method, the parameters have the following format
```
        roll,  pitch are in degrees
        yawrate is in degrees/s
        thrust is an integer value ranging from 10001 (next to no power) to 60000 (full power)
```
The values are packed into a radio packet that is sent to the Crazyflie and are received in the [`crtp_commander.c` file](https://github.com/bitcraze/crazyflie-firmware/blob/9343aa686600aa5f04beb43549c81bdfda1f9cb8/src/modules/src/crtp_commander.c#L115-L117). The data is unpacked and converted to a setpoint in [`crtpCommanderRpytDecodeSetpoint()`](https://github.com/bitcraze/crazyflie-firmware/blob/9343aa686600aa5f04beb43549c81bdfda1f9cb8/src/modules/src/crtp_commander_rpyt.c#L120-L224).

Here it becomes a bit complicated, there are lots of ifs and buts depending on modes and so on but you can check out the code to see what data that is actually set in the setpoint in your case. The setpoint structure is complex ([defined here](https://github.com/bitcraze/crazyflie-firmware/blob/9343aa686600aa5f04beb43549c81bdfda1f9cb8/src/modules/interface/stabilizer_types.h#L250-L271)) and supports multiple modes where some data might not be used. Similarly in the controllers, some controllers may not care about all values in the setpoint.

Anyhow, the setpoint that is generated is sent to the mellinger controller in the [`controllerMellinger()` function](https://github.com/bitcraze/crazyflie-firmware/blob/9343aa686600aa5f04beb43549c81bdfda1f9cb8/src/modules/src/controller/controller_mellinger.c#L115) and end up the lines you are interested in.

I think you need to dig into the code in [crtpCommanderRpytDecodeSetpoint()](https://github.com/bitcraze/crazyflie-firmware/blob/9343aa686600aa5f04beb43549c81bdfda1f9cb8/src/modules/src/crtp_commander_rpyt.c#L120-L224) to figure out exactly the values that are set in your case.
